### ISSUE TITLE
Unable to reproduce validation result using predict()/predict_prob() function

### ISSUE THREAD

**Issue by eulersformula** [Microsoft/LightGBM#437]
I am using a python version of LightGBM on a Mac OSX system and I am having some issues with prediction functions. In short I am not able to reproduce the validation result using predict()/predict_prob() right after calling fit() function. Here is the code:

clf = lgb.LGBMRegressor(boosting_type='gbdt', objective=objective, num_leaves=num_leaves, subsample=bagging_fraction, subsample_freq=1, colsample_bytree=feature_fraction, learning_rate=eta, n_estimators=nrounds, seed=seed, nthread=nthread, reg_lambda=1)

_ = clf.fit(dtrain, target[train_index], eval_metric=rmse, eval_set=[(dtrain, target[train_index]), (dtest, target[test_index])])

After this function finishes, I observed that 
[3533]	valid_0's error: 0.374629	**valid_1's error: 0.46246**

Then I immediately called 
preds = clf.predict(dtest)
print rmse(preds, target[test_index])

However this time, the error was shown to be 0.486488, which was quite a bit of difference than 0.46246. I have a similar issue with LGBMClassifier() function as well. Can anyone please look into this problem?

**Comment by guolinke** [comment]
@eulersformula 
I think the reason is the float accuracy by the multi-threading sum-up order. 
Can you try use small iterations and give me the differ ?

**Comment by eulersformula** [comment]
I retried using 300 iterations, and the difference is still quite large:

[300]	valid_0's error: 0.463488	**valid_1's error: 0.472815**
After using predict() function:
0.49015656486542553

**Comment by guolinke** [comment]
It is almost same in my test.
You use the custom eval function, right?
can you paste the code of ```rmse``` ?

**Comment by guolinke** [comment]
@wxchan can you also help to take a look for this issue ?

**Comment by eulersformula** [comment]
Yes I am using customized eval function. I defined it as 

def rmse(preds, labels):
    return 'error', np.sqrt(((np.clip(preds, 0, 2) - labels) ** 2).mean()), False

For your convenience I also pasted the several lines of relevant codes below:

clf = lgb.LGBMRegressor(boosting_type='gbdt', objective=objective, num_leaves=num_leaves, subsample=bagging_fraction, subsample_freq=1, colsample_bytree=feature_fraction, learning_rate=eta, n_estimators=nrounds, seed=seed, nthread=nthread, reg_lambda=1)
_ = clf.fit(dtrain, target[train_index], eval_metric=rmse, eval_set=[(dtrain, target[train_index]), (dtest, target[test_index])])
preds = clf.predict(dtest, num_iteration=nrounds)
print rmse(preds, target[test_index])

The final score was generated by the last line.

**Comment by wxchan** [comment]
self-defined eval function should be `func(y_true, y_pred)`, you are using `func(y_pred, y_true)`; I guess it's the reason

**Comment by eulersformula** [comment]
I redefined it as following:

def rmse(labels, preds):
    return 'error', np.sqrt(((preds - labels) ** 2).mean()), False

But I still got a huge different (same as before). Actually since the function is symmetric about labels and predictions, it would not make a difference.

Do you guys have any bench mark data and script such that I could try to replicate the results? I tried to use single thread instead of 4 and that didn't help.

**Comment by wxchan** [comment]
is your predict result in any iteration of training log?

**Comment by wxchan** [comment]
you might use datasets in examples folder

**Comment by eulersformula** [comment]
OK... I think I finally figured out what's happening here. Basically I have to fill NA's in the data before running the whole thing. Interestingly, without filling in NA's in the data, the training process can still be performed and the rmse of the validation set is actually correct! However when using predict() function to the same validation set, it has problems. I am not sure if this is a bug, but just to quickly mention that xgboost supports self-handling of the NA's without problems. To clarify, I am not saying that xgboost is better than LightGBM - actually thanks for writing this great tool which is way faster than xgboost!

**Comment by guolinke** [comment]
@eulersformula 
yea, we don't support the na input in python now.
I will quick fix for that.

**Comment by Frank-AFN** [comment]
Does the problem fix? I also come across the same issues when using scikit-learn wrapper LGBMRegressor. can not reproduce the results. I also can not use random_state param when to create a model.

**Comment by guolinke** [comment]
@Frank-AFN 
it is fixed, please try the latest code.
if not, please provide a re-produce example.

**Comment by Frank-AFN** [comment]
class lightgbm.LGBMRegressor(boosting_type='gbdt', num_leaves=31, max_depth=-1, learning_rate=0.1, n_estimators=10, max_bin=255, subsample_for_bin=50000, objective=None, min_split_gain=0, min_child_weight=5, min_child_samples=10, subsample=1, subsample_freq=1, colsample_bytree=1, reg_alpha=0, reg_lambda=0, random_state=0, n_jobs=-1, silent=True, **kwargs) 
As the doc http://lightgbm.readthedocs.io/en/latest/python/lightgbm.html said. I still can not use random_state param.. . so how can I add this param? @guolinke

**Comment by guolinke** [comment]
@wxchan

**Comment by Frank-AFN** [comment]
Hi @guolinke, maybe I have the same confusion with this guy. Same script and same data with preprocessing can not get the same score. Thanks for your reply.
https://www.kaggle.com/c/zillow-prize-1/discussion/37445

**Comment by guolinke** [comment]
@Frank-AFN 
when random seed is fixed, the result should be re-producable

**Comment by wxchan** [comment]
@Frank-AFN I just check latest codes, the result can be reproduced with same random_state. Can you share your whole codes?

**Comment by Frank-AFN** [comment]
@wxchan 
def cross_validation(cross_validation_results_file, best_model_file):
    x_data, y_data, train_columns = prepare_lgb_data()
    lgb_model = lgb.LGBMRegressor(nthread=48, seed=0)
    print(lgb_model.get_params().keys())
    print('train data column length is {0}'.format(len(train_columns)))

    cv_param = {'num_leaves': [512],
                'max_bin': [10],
                'boosting_type': ['gbdt'],
                'min_data': [600],
                'min_hessian': [0.05],
                'feature_fraction_seed': [2],
                'bagging_seed': [3],
                'sub_feature': [0.3],
                'bagging_fraction': [0.85],
                'bagging_freq': [40],
                'learning_rate': [0.01],
                'silent': [1],
                'n_estimators': [1000],
                'num_iterations': [1000],
                'max_depth': [8],
                'objective': ['regression_l1']}
    cv = KFold(n_splits=3, shuffle=False, random_state=0)

    clf = GridSearchCV(lgb_model,
                       cv_param,
                       n_jobs=-1,
                       cv=cv,
                       verbose=1,
                       scoring='neg_mean_absolute_error')
    clf.fit(x_data, y_data)

Could you tell me add the random_state param to which place?
Thanks~

**Comment by Frank-AFN** [comment]
@guolinke @wxchan

**Comment by wxchan** [comment]
you can add into cv_param

**Comment by Frank-AFN** [comment]
Still can not.  @wxchan

**Comment by wxchan** [comment]
@Frank-AFN can you paste your result here?
